usr/src/uts/common/fs/zfs/dsl_dataset.c

```diff
@@ -2501,11 +2501,38 @@ dsl_dataset_rollback_check(void *arg, dmu_tx_t *tx)
 	 * the latest snapshot is it.
 	 */
 	if (ddra->ddra_tosnap != NULL) {
-		char namebuf[ZFS_MAX_DATASET_NAME_LEN];
+		dsl_dataset_t *snapds;
 
-		dsl_dataset_name(ds->ds_prev, namebuf);
-		if (strcmp(namebuf, ddra->ddra_tosnap) != 0)
-			return (SET_ERROR(EXDEV));
+		/* Check if the target snapshot exists at all. */
+		error = dsl_dataset_hold(dp, ddra->ddra_tosnap, FTAG, &snapds);
+		if (error != 0) {
+			/* Distinguish between missing dataset and snapshot. */
```

Comment:
I think this is squashing `missing dataset and missing snapshot` to the same error, not distinguishing them.

Reply:
By "dataset" here I meant the parent dataset of the snapshot.  If it does not exists, then we pass up ENOENT, but if the snapshot does not exist, then we convert ENOENT to ESRCH.
I agree that the comment is confusing.  I'll try to write a more verbose one.